### PR TITLE
Add test cases for spaces between attributes

### LINKED ISSUE #1504: `BasicFormat` doesn't add spaces between attributes
### Description

When using `AttributeListBuilder`, the build result (formatted by `BasicFormat`) doesn't have spaces between each attributes, which is ill-formed in Swift.

### Steps to Reproduce

```swift
let builder = AttributeListSyntax {
  AttributeSyntax("@inlinable")
  AttributeSyntax("@discardableResult")
}
print(builder) // @inlinable@discardableResult
```

### PR DESCRIPTION
**Pull request by kimdv**
Add test cases for #1504

### CODE REVIEW

File: Sources/SwiftBasicFormat/generated/BasicFormat.swift [hunk 1]
```diff
@@ -245,6 +247,8 @@ open class BasicFormat: SyntaxRewriter {
      (.postfixQuestionMark, .rightAngle), // Ensures there is not a space in `ContiguousArray<RawSyntax?>`
      (.postfixQuestionMark, .rightParen): // Ensures there is not a space in `myOptionalClosure?()`
       return false
+    case (.identifier, .keyword):
```
Comment (stevapple) [inline Sources/SwiftBasicFormat/generated/BasicFormat.swift:250]: Add a comment explaining what it’s going to solve?

File: CodeGeneration/Sources/SyntaxSupport/AttributeNodes.swift [hunk 2]
```diff
@@ -19,7 +19,8 @@ public let ATTRIBUTE_NODES: [Node] = [
     element: "Syntax",
     elementName: "Attribute",
     elementChoices: ["Attribute", "IfConfigDecl"],
-    omitWhenEmpty: true
+    omitWhenEmpty: true,
+    elementsSeparatedByNewline: true
```
Comment (stevapple) [inline CodeGeneration/Sources/SyntaxSupport/AttributeNodes.swift:23]: We may need `elementsSeparatedBySpace` instead… Splitting the attribute list with newlines looks terrible, especially when they’re applied to a declaration without leading newline trivia.

File: Tests/SwiftSyntaxBuilderTest/FunctionTests.swift [hunk 3]
```diff
@@ -163,6 +163,45 @@ final class FunctionTests: XCTestCase {
...
+        """
+        @discardableResult public func foo() -> String {
+            return "foo"
+        }
+        """
+      ),
+      #line: (
+        DeclSyntax(
+            FunctionDeclSyntax(
+                attributes: [.attribute(AttributeSyntax("@discardableResult"))],
```
Comment (stevapple) [inline Tests/SwiftSyntaxBuilderTest/FunctionTests.swift:183]: Try multiple attributes here?

**Review by stevapple — CHANGES_REQUESTED** [review]

**Comment by kimdv** [comment]
Do you think we should add newline or a whitespace between @stevapple?

**Comment by stevapple** [comment]
I would prefer whitespaces to newlines for wherever it's applicable (aka. legal in Swift).

**Comment by kimdv** [comment]
I've been looking into some swift code and cannot conclude what is the best practice to do or if the `BasicFormat` should support it. 
IMO a newline is fine so it will generate valid swift and we can use what we already have. 

Then we can also avoid having the special handling `case (.identifier, .keyword):` inside `BasicFormat`

What do you think of that?

**Comment by stevapple** [comment]
Newlines are not very bad, but the current implementation will produce
```swift
@inlinable
@discardableResult
@available(*, deprecated, message: "Use print directly instead") public func hello() -> Int {
    print("Hello world!")
    return 0
}
```
which looks really weird.

The main reason to prefer whitespace is that, when we want to a newline we can always add `.with(\.trailingTrivia, .newline)`, but if we default to newline the reverse just won’t work.

**Comment by stevapple** [comment]
Separating attributes with newlines maybe more troublesome than it seems… After some investigation.

```swift

let buildable = FunctionDeclSyntax(
  identifier: "hello",
  signature: .init(
    input: .init(
      parameterList: [
        FunctionParameterSyntax(
          attributes: [.attribute("@NameBuilder")],
          firstName: "name",
          type: AttributedTypeSyntax(
            attributes: [.attribute("@escaping"), .attribute("@Sendable")],
            baseType: FunctionTypeSyntax(
              arguments: [],
              output: .init(returnType: TypeSyntax("String"))
            )
          )
        )
      ]
    )
  ),
  body: .init(
    statements: [
      .init(item: .decl(#"print("Hello, \(name())!")"#))
    ]
  )
)
print(buildable.formatted())
```

Gives:
```swift
func hello(
    @NameBuilder
    name: 
    @escaping
    @Sendable
    () -> String) {print("Hello, \(name())!")
}
```

Which is, IMO, valid but almost unacceptable.

**Comment by kimdv** [comment]
I see.. hmm.
Then we either should handle case by case or only add white space and let users add new line.

**Comment by stevapple** [comment]
I would suggest the latter, leaving case-by-case formatting for `swift-format`.

**Comment by ahoppen** [comment]
> At least we should have a way to opt out of these new lines, which I didn’t see at the moment.

Just for the record, I don’t think we should add many customization options to `BasicFormat`. It is called `BasicFormat` for a reason: It’s goal is to format a SwiftSyntax tree in a way that, most importantly, parses and, secondly, doesn’t look terrible. I am fine with requiring users to subclass `BasicFormat` if they want to customize it.

**Comment by stevapple** [comment]
@ahoppen I didn’t meant to add the option to `BasicFormat` actually. In existing APIs we already can use explicit newlines over spaces by setting the trivia manually, without touching `BasicFormat` at all.

You can see my use case in https://github.com/teco-project/teco-code-generators, which is used to generate ~1 million lines of nice-looking code with `SwiftSyntaxBuilder`, with little modification to `BasicFormat`.

**Comment by kimdv** [comment]
@ahoppen changed this to add some test cases that for #1504

**Comment by kimdv** [comment]
@swift-ci please test

File: Tests/SwiftSyntaxBuilderTest/SourceFileTests.swift [hunk 4]
```diff
@@ -31,13 +31,12 @@ final class SourceFileTests: XCTestCase {
     assertBuildResult(
       source,
       """
-      import Foundation
-      import UIKit
+      import Foundation 
+      import UIKit
```
Comment (stevapple) [inline Tests/SwiftSyntaxBuilderTest/SourceFileTests.swift:35]: Looks like a regression here. Is this intentional?

**Review by stevapple — APPROVED** [review]
Looks good overall!

**Comment by stevapple** [comment]
Should there be test cases for #1472 and #1551 too?

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test windows

File: Tests/SwiftSyntaxBuilderTest/TriviaTests.swift [hunk 5]
```diff
@@ -83,12 +83,14 @@ final class TriviaTests: XCTestCase {
           AttributeSyntax(attributeName: TypeSyntax("A")).with(\.trailingTrivia, .space)
           AttributeSyntax(attributeName: TypeSyntax("B")).with(\.trailingTrivia, .space)
         },
-        "@A @B "
+        """
+        @A @B
```
Comment (ahoppen) [inline Tests/SwiftSyntaxBuilderTest/TriviaTests.swift:87]: I would prefer this string literal to continue being single line so you can actually see the trailing whitespace.

**Review by ahoppen — APPROVED** [review]

**Comment by kimdv** [comment]
@swift-ci please test

**Pull request merged**